services/worker-manager/src/providers/gcpprovider.js

```diff
@@ -0,0 +1,175 @@
+'use strict';
+
+const {Bid} = require('../bid');
+const {Provider} = require('../provider');
+const Worker = require('../worker');
+const Compute = require('@google-cloud/compute');
+
+class GCPProvider extends Provider {
+
+  /**
+   * Construct a GCP Provider
+   */
+  constructor({id}) {
+    super({id});
+    this.gcpCompute = new Compute();
+  }
+
+  convertToGCPStatus(taskclusterWorkerState) {
+    switch (taskclusterWorkerState) {
+      case Provider.states.booting: return ['STAGING'];
+      case Provider.states.requested: return ['PROVISIONING'];
+      case Provider.states.running: return ['RUNNING'];
+      case Provider.states.terminating: return ['STOPPING', 'SUSPENDING', 'SUSPENDED', 'TERMINATED'];
+    }
+  }
+
+  convertFromGCPStatus(gcpInstanceStatus) {
+    switch (gcpInstanceStatus) {
+      case 'PROVISIONING': return Provider.states.requested;
+      case 'STAGING': return Provider.states.requested;
+      case 'RUNNING': return Provider.states.running;
+      case 'STOPPING': return Provider.states.terminating;
+      case 'SUSPENDING': return Provider.states.terminating;
+      case 'SUSPENDED': return Provider.states.terminating;
+      case 'TERMINATED': return Provider.states.terminating;
+    }
+  }
+
+  /**
+   * this is called in provisioner
+   *
+   * @param states Array<String>
+   * @param workerTypes Array<String>
+   * @returns {Promise<Array<Worker>>}
+   */
+  async listWorkers({states, workerTypes}) {
+    console.log(`Got states: ${states}, and worker types: ${workerTypes}`);
+
+    let filter = workerTypes.map(wt => `(labels.worker-type = ${wt})`).join(' OR ');
```

Comment:
Is it possible to include the states in this filter we send to gcp as well?

Reply:
Short answer is no

As you can see, the filter is conveniently a string with a rather simplistic syntax, it doesn't seem to be created for anything more complicated that couple of conditions. I tried experimenting, but it either complains on the syntax or doesn't work. 

Also, putting together that string is obnoxious, hard to write, read and debug. I decided to compromise: request filtering by worker type will produce less results than by status; and then iterate over and filter by status.

Reply:
Actually, I decided to make sure I'm not wrong, and turns out it's actually possible